// @codingame/monaco-vscode-7bbc9e7d-eeae-55fc-8bf9-dc2f66e0dc73-common@14.0.6/vscode/src/vs/workbench/contrib/terminalContrib/typeAhead/common/terminalTypeAheadConfiguration.js

import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';

const DEFAULT_LOCAL_ECHO_EXCLUDE = ['vim', 'vi', 'nano', 'tmux'];
var TerminalTypeAheadSettingId;
(function (TerminalTypeAheadSettingId) {
    TerminalTypeAheadSettingId["LocalEchoLatencyThreshold"] = "terminal.integrated.localEchoLatencyThreshold";
    TerminalTypeAheadSettingId["LocalEchoEnabled"] = "terminal.integrated.localEchoEnabled";
    TerminalTypeAheadSettingId["LocalEchoExcludePrograms"] = "terminal.integrated.localEchoExcludePrograms";
    TerminalTypeAheadSettingId["LocalEchoStyle"] = "terminal.integrated.localEchoStyle";
})(TerminalTypeAheadSettingId || (TerminalTypeAheadSettingId = {}));
const terminalTypeAheadConfiguration = {
    [TerminalTypeAheadSettingId.LocalEchoLatencyThreshold]: {
        description: ( localize(
            10185,
            "Length of network delay, in milliseconds, where local edits will be echoed on the terminal without waiting for server acknowledgement. If '0', local echo will always be on, and if '-1' it will be disabled."
        )),
        type: 'integer',
        minimum: -1,
        default: 30,
    },
    [TerminalTypeAheadSettingId.LocalEchoEnabled]: {
        markdownDescription: ( localize(
            10186,
            "When local echo should be enabled. This will override {0}",
            '`#terminal.integrated.localEchoLatencyThreshold#`'
        )),
        type: 'string',
        enum: ['on', 'off', 'auto'],
        enumDescriptions: [
            ( localize(10187, "Always enabled")),
            ( localize(10188, "Always disabled")),
            ( localize(10189, "Enabled only for remote workspaces"))
        ],
        default: 'auto'
    },
    [TerminalTypeAheadSettingId.LocalEchoExcludePrograms]: {
        description: ( localize(
            10190,
            "Local echo will be disabled when any of these program names are found in the terminal title."
        )),
        type: 'array',
        items: {
            type: 'string',
            uniqueItems: true
        },
        default: DEFAULT_LOCAL_ECHO_EXCLUDE,
    },
    [TerminalTypeAheadSettingId.LocalEchoStyle]: {
        description: ( localize(
            10191,
            "Terminal style of locally echoed text; either a font style or an RGB color."
        )),
        default: 'dim',
        anyOf: [
            {
                enum: ['bold', 'dim', 'italic', 'underlined', 'inverted', '#ff0000'],
            },
            {
                type: 'string',
                format: 'color-hex',
            }
        ]
    },
};

export { DEFAULT_LOCAL_ECHO_EXCLUDE, TerminalTypeAheadSettingId, terminalTypeAheadConfiguration };
