// @codingame/monaco-vscode-api@14.0.6/vscode/src/vs/platform/contextkey/common/contextkeys.js

import { isMacintosh, isLinux, isWindows, isWeb, isIOS, isMobile } from '../../../base/common/platform.js';
import { localize } from '../../../nls.js';
import { RawContextKey } from './contextkey.js';

const IsMacContext = ( new RawContextKey('isMac', isMacintosh, ( localize(1675, "Whether the operating system is macOS"))));
const IsLinuxContext = ( new RawContextKey('isLinux', isLinux, ( localize(1676, "Whether the operating system is Linux"))));
const IsWindowsContext = ( new RawContextKey('isWindows', isWindows, ( localize(1677, "Whether the operating system is Windows"))));
const IsWebContext = ( new RawContextKey('isWeb', isWeb, ( localize(1678, "Whether the platform is a web browser"))));
const IsMacNativeContext = ( new RawContextKey('isMacNative', isMacintosh && !isWeb, ( localize(1679, "Whether the operating system is macOS on a non-browser platform"))));
const IsIOSContext = ( new RawContextKey('isIOS', isIOS, ( localize(1680, "Whether the operating system is iOS"))));
const IsMobileContext = ( new RawContextKey('isMobile', isMobile, ( localize(1681, "Whether the platform is a mobile web browser"))));
const IsDevelopmentContext = ( new RawContextKey('isDevelopment', false, true));
const ProductQualityContext = ( new RawContextKey('productQualityType', '', ( localize(1682, "Quality type of VS Code"))));
const InputFocusedContextKey = 'inputFocus';
const InputFocusedContext = ( new RawContextKey(InputFocusedContextKey, false, ( localize(1683, "Whether keyboard focus is inside an input box"))));

export { InputFocusedContext, InputFocusedContextKey, IsDevelopmentContext, IsIOSContext, IsLinuxContext, IsMacContext, IsMacNativeContext, IsMobileContext, IsWebContext, IsWindowsContext, ProductQualityContext };
