// @codingame/monaco-vscode-api@14.0.6/vscode/src/vs/editor/contrib/inlineCompletions/browser/view/inlineEdits/gutterIndicatorMenu.js

import { __decorate, __param } from '../../../../../../../../../external/tslib/tslib.es6.js';
import { renderIcon } from '../../../../../../base/browser/ui/iconLabel/iconLabels.js';
import { KeybindingLabel, unthemedKeybindingLabelOptions } from '../../../../../../base/browser/ui/keybindingLabel/keybindingLabel.js';
import { Codicon } from '../../../../../../base/common/codicons.js';
import { observableValue } from '../../../../../../base/common/observableInternal/base.js';
import '../../../../../../base/common/arrays.js';
import '../../../../../../base/common/event.js';
import '../../../../../../base/common/lifecycle.js';
import { autorun } from '../../../../../../base/common/observableInternal/autorun.js';
import { derived, derivedWithStore } from '../../../../../../base/common/observableInternal/derived.js';
import '../../../../../../base/common/cancellation.js';
import { constObservable, observableFromEvent } from '../../../../../../base/common/observableInternal/utils.js';
import { OS } from '../../../../../../base/common/platform.js';
import { ThemeIcon } from '../../../../../../base/common/themables.js';
import { localize } from '../../../../../../nls.js';
import { ICommandService } from '../../../../../../platform/commands/common/commands.service.js';
import { IContextKeyService } from '../../../../../../platform/contextkey/common/contextkey.service.js';
import { IKeybindingService } from '../../../../../../platform/keybinding/common/keybinding.service.js';
import { AcceptInlineCompletion, JumpToNextInlineEdit, HideInlineCompletion } from '../../controller/commands.js';
import { n } from './utils.js';

let GutterIndicatorMenuContent = class GutterIndicatorMenuContent {
    constructor(_menuTitle, _tabAction, _close, _extensionCommands, _contextKeyService, _keybindingService, _commandService) {
        this._menuTitle = _menuTitle;
        this._tabAction = _tabAction;
        this._close = _close;
        this._extensionCommands = _extensionCommands;
        this._contextKeyService = _contextKeyService;
        this._keybindingService = _keybindingService;
        this._commandService = _commandService;
    }
    toDisposableLiveElement() {
        return this._createHoverContent().toDisposableLiveElement();
    }
    _createHoverContent() {
        const activeElement = observableValue('active', undefined);
        const createOptionArgs = (options) => {
            return {
                title: options.title,
                icon: options.icon,
                keybinding: typeof options.commandId === 'string' ? this._getKeybinding(options.commandArgs ? undefined : options.commandId) : derived(reader => typeof options.commandId === 'string' ? undefined : this._getKeybinding(options.commandArgs ? undefined : options.commandId.read(reader)).read(reader)),
                isActive: ( activeElement.map(v => v === options.id)),
                onHoverChange: v => activeElement.set(v ? options.id : undefined, undefined),
                onAction: () => {
                    this._close(true);
                    return this._commandService.executeCommand(typeof options.commandId === 'string' ? options.commandId : options.commandId.get(), ...(options.commandArgs ?? []));
                },
            };
        };
        return hoverContent([
            header(this._menuTitle),
            option(createOptionArgs({ id: 'gotoAndAccept', title: `${( localize(1232, "Go To"))} / ${( localize(1233, "Accept"))}`, icon: ( this._tabAction.map(action => action === 'accept' ? Codicon.check : Codicon.arrowRight)), commandId: ( this._tabAction.map(action => action === 'accept' ? ( new AcceptInlineCompletion()).id : ( new JumpToNextInlineEdit()).id)) })),
            option(createOptionArgs({ id: 'reject', title: ( localize(1234, "Reject")), icon: Codicon.close, commandId: ( new HideInlineCompletion()).id })),
            separator(),
            this._extensionCommands?.map(c => c && c.length > 0 ? [
                ...( c.map(
                    c => option(createOptionArgs({ id: c.id, title: c.title, icon: Codicon.symbolEvent, commandId: c.id, commandArgs: c.arguments }))
                )),
                separator()
            ] : []),
            option(createOptionArgs({ id: 'settings', title: ( localize(1235, "Settings")), icon: Codicon.gear, commandId: 'workbench.action.openSettings', commandArgs: ['@tag:nextEditSuggestions'] })),
        ]);
    }
    _getKeybinding(commandId) {
        if (!commandId) {
            return constObservable(undefined);
        }
        return observableFromEvent(this._contextKeyService.onDidChangeContext, () => this._keybindingService.lookupKeybinding(commandId));
    }
};
GutterIndicatorMenuContent = ( __decorate([
    ( __param(4, IContextKeyService)),
    ( __param(5, IKeybindingService)),
    ( __param(6, ICommandService))
], GutterIndicatorMenuContent));
function hoverContent(content) {
    return n.div({
        class: 'content',
        style: {
            margin: 4,
            minWidth: 150,
        }
    }, content);
}
function header(title) {
    return n.div({
        class: 'header',
        style: {
            color: 'var(--vscode-descriptionForeground)',
            fontSize: '12px',
            fontWeight: '600',
            padding: '0 10px',
            lineHeight: 26,
        }
    }, [title]);
}
function option(props) {
    return derivedWithStore((_reader, store) => n.div({
        class: ['monaco-menu-option', props.isActive?.map(v => v && 'active')],
        onmouseenter: () => props.onHoverChange?.(true),
        onmouseleave: () => props.onHoverChange?.(false),
        onclick: props.onAction,
        onkeydown: e => {
            if (e.key === 'Enter') {
                props.onAction?.();
            }
        },
        tabIndex: 0,
    }, [
        n.elem('span', {
            style: {
                fontSize: 16,
                display: 'flex',
            }
        }, [ThemeIcon.isThemeIcon(props.icon) ? renderIcon(props.icon) : ( props.icon.map(icon => renderIcon(icon)))]),
        n.elem('span', {}, [props.title]),
        n.div({
            style: { marginLeft: 'auto', opacity: '0.6' },
            ref: elem => {
                const keybindingLabel = store.add(( new KeybindingLabel(elem, OS, { disableTitle: true, ...unthemedKeybindingLabelOptions })));
                store.add(autorun(reader => {
                    keybindingLabel.set(props.keybinding.read(reader));
                }));
            }
        })
    ]));
}
function separator() {
    return n.div({
        class: 'menu-separator',
        style: {
            color: 'var(--vscode-editorActionList-foreground)',
            padding: '2px 0',
        }
    }, n.div({
        style: {
            borderBottom: '1px solid var(--vscode-editorHoverWidget-border)',
        }
    }));
}

export { GutterIndicatorMenuContent };
