// @codingame/monaco-vscode-7bbc9e7d-eeae-55fc-8bf9-dc2f66e0dc73-common@14.0.6/vscode/src/vs/workbench/contrib/terminalContrib/suggest/common/terminalSuggestConfiguration.js

import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';
import { TerminalSettingId } from '@codingame/monaco-vscode-api/vscode/vs/platform/terminal/common/terminal';

var TerminalSuggestSettingId;
(function (TerminalSuggestSettingId) {
    TerminalSuggestSettingId["Enabled"] = "terminal.integrated.suggest.enabled";
    TerminalSuggestSettingId["QuickSuggestions"] = "terminal.integrated.suggest.quickSuggestions";
    TerminalSuggestSettingId["SuggestOnTriggerCharacters"] = "terminal.integrated.suggest.suggestOnTriggerCharacters";
    TerminalSuggestSettingId["RunOnEnter"] = "terminal.integrated.suggest.runOnEnter";
    TerminalSuggestSettingId["BuiltinCompletions"] = "terminal.integrated.suggest.builtinCompletions";
    TerminalSuggestSettingId["WindowsExecutableExtensions"] = "terminal.integrated.suggest.windowsExecutableExtensions";
    TerminalSuggestSettingId["Providers"] = "terminal.integrated.suggest.providers";
    TerminalSuggestSettingId["ShowStatusBar"] = "terminal.integrated.suggest.showStatusBar";
})(TerminalSuggestSettingId || (TerminalSuggestSettingId = {}));
const windowsDefaultExecutableExtensions = [
    'exe',
    'bat',
    'cmd',
    'com',
    'msi',
    'ps1',
    'vbs',
    'js',
    'jar',
    'py',
    'rb',
    'pl',
    'sh',
];
const terminalSuggestConfigSection = 'terminal.integrated.suggest';
const terminalSuggestConfiguration = {
    [TerminalSuggestSettingId.Enabled]: {
        restricted: true,
        markdownDescription: ( localize(
            10170,
            "Enables experimental terminal Intellisense suggestions for supported shells ({0}) when {1} is set to {2}.\n\nIf shell integration is installed manually, {3} needs to be set to {4} before calling the shell integration script.",
            'PowerShell v7+, zsh, bash, fish',
            `\`#${TerminalSettingId.ShellIntegrationEnabled}#\``,
            '`true`',
            '`VSCODE_SUGGEST`',
            '`1`'
        )),
        type: 'boolean',
        default: false,
        tags: ['experimental'],
    },
    [TerminalSuggestSettingId.Providers]: {
        restricted: true,
        markdownDescription: ( localize(
            10171,
            "Providers are enabled by default. Omit them by setting the id of the provider to `false`."
        )),
        type: 'object',
        properties: {},
        default: {
            'terminal-suggest': true,
            'pwsh-shell-integration': false,
        },
        tags: ['experimental'],
    },
    [TerminalSuggestSettingId.QuickSuggestions]: {
        restricted: true,
        markdownDescription: ( localize(
            10172,
            "Controls whether suggestions should automatically show up while typing. Also be aware of the {0}-setting which controls if suggestions are triggered by special characters.",
            `\`#${TerminalSuggestSettingId.SuggestOnTriggerCharacters}#\``
        )),
        type: 'boolean',
        default: true,
    },
    [TerminalSuggestSettingId.SuggestOnTriggerCharacters]: {
        restricted: true,
        markdownDescription: ( localize(
            10173,
            "Controls whether suggestions should automatically show up when typing trigger characters."
        )),
        type: 'boolean',
        default: true,
    },
    [TerminalSuggestSettingId.RunOnEnter]: {
        restricted: true,
        markdownDescription: ( localize(
            10174,
            "Controls whether suggestions should run immediately when `Enter` (not `Tab`) is used to accept the result."
        )),
        enum: ['ignore', 'never', 'exactMatch', 'exactMatchIgnoreExtension', 'always'],
        markdownEnumDescriptions: [
            ( localize(
            10175,
            "Ignore suggestions and send the enter directly to the shell without completing. This is used as the default value so the suggest widget is as unobtrusive as possible."
        )),
            ( localize(10176, "Never run on `Enter`.")),
            ( localize(10177, "Run on `Enter` when the suggestion is typed in its entirety.")),
            ( localize(
            10178,
            "Run on `Enter` when the suggestion is typed in its entirety or when a file is typed without its extension included."
        )),
            ( localize(10179, "Always run on `Enter`."))
        ],
        default: 'ignore',
    },
    [TerminalSuggestSettingId.BuiltinCompletions]: {
        restricted: true,
        markdownDescription: ( localize(
            10180,
            "Controls which built-in completions are activated. This setting can cause conflicts if custom shell completions are configured in the shell profile."
        )),
        type: 'object',
        properties: {
            'pwshCode': {
                description: ( localize(
                    10181,
                    'Custom PowerShell argument completers will be registered for VS Code\'s `code` and `code-insiders` CLIs. This is currently very basic and always suggests flags and subcommands without checking context.'
                )),
                type: 'boolean'
            },
            'pwshGit': {
                description: ( localize(
                    10182,
                    'Custom PowerShell argument completers will be registered for the `git` CLI.'
                )),
                type: 'boolean'
            },
        },
        default: {
            pwshCode: true,
            pwshGit: true,
        }
    },
    [TerminalSuggestSettingId.WindowsExecutableExtensions]: {
        restricted: true,
        markdownDescription: ( localize(
            10183,
            "A set of windows command executable extensions that will be included as suggestions in the terminal.\n\nMany executables are included by default, listed below:\n\n{0}.\n\nTo exclude an extension, set it to `false`\n\n. To include one not in the list, add it and set it to `true`.",
            ( windowsDefaultExecutableExtensions.sort().map(extension => `- ${extension}`)).join('\n')
        )),
        type: 'object',
        default: {},
        tags: ['experimental'],
    },
    [TerminalSuggestSettingId.ShowStatusBar]: {
        restricted: true,
        markdownDescription: ( localize(
            10184,
            "Controls whether the terminal suggestions status bar should be shown."
        )),
        type: 'boolean',
        default: true,
        tags: ['experimental'],
    },
};

export { TerminalSuggestSettingId, terminalSuggestConfigSection, terminalSuggestConfiguration, windowsDefaultExecutableExtensions };
